safekeeper/src/metrics.rs

```diff
@@ -63,13 +66,120 @@ pub static PERSIST_CONTROL_FILE_SECONDS: Lazy<Histogram> = Lazy::new(|| {
 });
 pub static PG_IO_BYTES: Lazy<IntCounterVec> = Lazy::new(|| {
     register_int_counter_vec!(
-        "safekeeper_pg_io_bytes",
+        "safekeeper_pg_io_bytes_total",
         "Bytes read from or written to any PostgreSQL connection",
-        &["direction"]
+        &["client_az", "sk_az", "app_name", "dir", "same_az"]
     )
     .expect("Failed to register safekeeper_pg_io_bytes gauge")
 });
 
+pub const LABEL_UNKNOWN: &str = "unknown";
+
+/// Labels for traffic metrics.
+struct ConnectionLabels {
+    /// Availability zone of the connection origin.
+    client_az: String,
+    /// Availability zone of the current safekeeper.
+    sk_az: String,
+    /// Client application name.
+    app_name: String,
+    /// Total bytes read from this connection.
+    read: GenericCounter<metrics::core::AtomicU64>,
+    /// Total bytes written to this connection.
+    write: GenericCounter<metrics::core::AtomicU64>,
+}
+
+impl ConnectionLabels {
+    fn new() -> Self {
+        let mut this = Self {
+            client_az: LABEL_UNKNOWN.to_string(),
+            sk_az: LABEL_UNKNOWN.to_string(),
+            app_name: LABEL_UNKNOWN.to_string(),
+            // Fill in the metrics with dummy values, they will be updated later.
+            read: GenericCounter::new("_", "_").unwrap(),
+            write: GenericCounter::new("_", "_").unwrap(),
+        };
+        this.update_metrics();
+        this
+    }
+
+    fn update_metrics(&mut self) {
+        let same_az = match (self.client_az.as_str(), self.sk_az.as_str()) {
+            (LABEL_UNKNOWN, _) | (_, LABEL_UNKNOWN) => LABEL_UNKNOWN,
+            (client_az, sk_az) => {
+                if client_az == sk_az {
+                    "true"
+                } else {
+                    "false"
+                }
+            }
+        };
+
+        self.read = PG_IO_BYTES.with_label_values(&[
+            &self.client_az,
+            &self.sk_az,
+            &self.app_name,
+            "read",
+            same_az,
+        ]);
+        self.write = PG_IO_BYTES.with_label_values(&[
+            &self.client_az,
+            &self.sk_az,
+            &self.app_name,
+            "write",
+            same_az,
+        ]);
+    }
+}
+
+/// Metrics for measuring traffic (r/w bytes) in a single PostgreSQL connection.
+#[derive(Clone)]
+pub struct TrafficMetrics {
+    labels: Arc<RwLock<ConnectionLabels>>,
```

Comment:
Not sure this is important, but having here completely unneeded rwlock is indeed not appealing. I haven't thought too deep, but I'd try Refcell here, conceptually if fits here. However, I suspect it'd fail due to async code: futures should be Send, and Arc is Send iff underlying type is Sync, and Refcell is not. But better try. In any case worth a comment.

Reply:
Yes, `RefCell` fails on the line with `block_on(.. pgbackend.run ..)`:

```
`RefCell<TrafficMetricsState>` cannot be shared between threads safely
the trait `std::marker::Sync` is not implemented for `RefCell<TrafficMetricsState>`
the trait `Handler<IO>` is implemented for `SafekeeperPostgresHandler`
required for `Arc<RefCell<TrafficMetricsState>>` to implement `std::marker::Sync`
required for `&TrafficMetrics` to implement `Send`
```

I'd say that `RwLock` is not that bad, we acquire a write lock only once (on startup packet), and all subsequent read locks will only increase/decrease internal atomic read counter, that shouldn't be expensive.

But I can try to use `AtomicPtr`, perhaps it's possible to use it here?

Reply:
Ah, I didn't realize atomic counters modifications don't need mut.

> But I can try to use AtomicPtr, perhaps it's possible to use it here?
Not immediately sure. Seems like `Arc<AtomicPtr<ConnectionLabels>` is probably not good because AtomicPtr doesn't own the data. Anyway it is also ugly.

I recently wrote SendRefcell specifically for this case (we need it to be Send only to satisfy Future bounds), I can revive that. But probably it is not worth it atm.

Reply:
I searched for a ready alternatives a bit more:
- `ShardedLock`, same as `RwLock`, but with faster reads  – https://docs.rs/crossbeam/latest/crossbeam/sync/struct.ShardedLock.html
- `ArcSwap`, something like `Atomic<Arc<T>>`, to never block reads – https://docs.rs/arc-swap/latest/arc_swap/

I think `RwLock` will be ok for now, and later we can look at alternatives above.